circuit/src/native.rs

```diff
@@ -0,0 +1,403 @@
+
+use crate::eddsa::native::{PublicKey, Signature};
+use halo2::{
+	arithmetic::Field,
+	halo2curves::{bn256::Fr, FieldExt},
+};
+
+const NUM_NEIGHBOURS: usize = 6;
+const NUM_ITERATIONS: usize = 20;
+const INITIAL_SCORE: u128 = 1000;
+
+struct Opinion {
+	sig: Signature,
+	message_hash: Fr,
+	scores: [(PublicKey, Fr); NUM_NEIGHBOURS],
+}
+
+impl Opinion {
+	pub fn new(
+		sig: Signature, message_hash: Fr, scores: [(PublicKey, Fr); NUM_NEIGHBOURS],
+	) -> Self {
+		Self { sig, message_hash, scores }
+	}
+}
+
+impl Default for Opinion {
+	fn default() -> Self {
+		let sig = Signature::new(Fr::zero(), Fr::zero(), Fr::zero());
+		let message_hash = Fr::zero();
+		let scores = [(PublicKey::default(), Fr::zero()); NUM_NEIGHBOURS];
+		Self { sig, message_hash, scores }
+	}
+}
+
+struct EigenTrustSet {
+	set: [(PublicKey, Fr); NUM_NEIGHBOURS],
+	ops: HashMap<PublicKey, Opinion>,
+}
+
+impl EigenTrustSet {
+	pub fn new() -> Self {
+		Self { set: [(PublicKey::default(), Fr::zero()); NUM_NEIGHBOURS], ops: HashMap::new() }
+	}
+
+	pub fn add_member(&mut self, pk: PublicKey) {
+		let pos = self.set.iter().position(|&(x, _)| x == pk);
+		let first_available = self.set.iter().position(|&(x, _)| x == PublicKey::default());
+		// Make sure not already in the set
+		assert!(pos.is_none());
+
+		let initial_score = Fr::from_u128(INITIAL_SCORE);
+		let index = first_available.unwrap();
+		self.set[index] = (pk, initial_score);
+	}
+
+	pub fn update_op(&mut self, from: PublicKey, op: Opinion) {
+		let pos_from = self.set.iter().position(|&(x, _)| x == from);
+		assert!(pos_from.is_some());
+		self.ops.insert(from, op);
+	}
+
+	pub fn converge(&mut self) -> [Fr; NUM_NEIGHBOURS] {
+		let initial_score = Fr::from_u128(INITIAL_SCORE);
+
+		for i in 0..NUM_NEIGHBOURS {
+			let pk = self.set[i].0;
+
+			// Validity checks
+			if pk == PublicKey::default() {
+				assert!(self.ops.get_mut(&pk).is_none());
+			} else {
+				let mut ops_i = self.ops.get_mut(&pk).unwrap();
+
+				// Nullify scores that are given to wrong member at specific index
+				let mut score_sum = Fr::zero();
+				for j in 0..NUM_NEIGHBOURS {
+					let (pk_j, score) = ops_i.scores[j].clone();
+					let true_score =
+						if j != i && self.set[j].0 == pk_j { score } else { Fr::zero() };
+					ops_i.scores[j].1 = true_score;
+					score_sum += true_score;
+				}
+
+				// First we need filtered set -- set after we invalidated
+				// invalid peers:
+				// - Peers that dont have at least one valid score
+				if score_sum == Fr::zero() {
+					self.set[i] = (PublicKey::default(), Fr::zero());
+				} else {
+					// Normalize the scores
+					// Distribute the credits(INITIAL_SCORE) to the valid opinion values
+					// We assume that the initial credits of peer is constant(INITIAL_SCORE)
+					for j in 0..NUM_NEIGHBOURS {
+						let (_, op_score) = ops_i.scores[j].clone();
+						ops_i.scores[j].1 = op_score * score_sum.invert().unwrap() * initial_score;
+					}
+				}
+			}
+		}
```

Comment:
Can we make this more readable by making less logic branches?
Perhaps break into several loops instead of having just one.

Reply:
Yes, I done some refactoring in https://github.com/eigen-trust/protocol/pull/154/commits/0c365b1ffd58962ddf16563fc280a1391b3aca91